crash in raft.sendAppend: requested index at entry is unavailable

This happened in tandem with #1185 and after a flurry of messages

```
W0529 13:18:33.251543       1 dist_sender.go:588] failed to invoke InternalPushTxn: key range "00065201
"-"" outside of bounds of range ""-""
```

(which are not surprising in this scenario given that the node never executes any of the splits)

```
I0529 13:18:33.251657       1 retry.go:95] routing InternalPushTxn rpc failed; retrying immediately
W0529 13:18:33.251948       1 dist_sender.go:588] failed to invoke InternalPushTxn: key range "00065201"-"" outside of bounds of range ""-""
I0529 13:18:33.251966       1 retry.go:95] routing InternalPushTxn rpc failed; retrying immediately
W0529 13:18:33.253062       1 dist_sender.go:588] failed to invoke InternalPushTxn: key range "00065201"-"" outside of bounds of range ""-""
I0529 13:18:33.253138       1 retry.go:95] routing InternalPushTxn rpc failed; retrying immediately
panic: requested entry at index is unavailable

goroutine 64 [running]:
github.com/coreos/etcd/raft.(*raft).sendAppend(0xc2097f4000, 0x300000003)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/raft.go:250 +0x159
github.com/coreos/etcd/raft.stepLeader(0xc2097f4000, 0x4, 0x200000002, 0x300000003, 0x6, 0x0, 0xa, 0x0, 0x0, 0x0, ...)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/raft.go:523 +0x6f3
github.com/coreos/etcd/raft.(*raft).Step(0xc2097f4000, 0x4, 0x200000002, 0x300000003, 0x6, 0x0, 0xa, 0x0, 0x0, 0x0, ...)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/raft.go:487 +0x259
github.com/coreos/etcd/raft.(*multiNode).run(0xc2080ff0e0)
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/multinode.go:236 +0xd0b
created by github.com/coreos/etcd/raft.StartMultiNode
        /home/tobias/Code/Go/src/github.com/coreos/etcd/raft/multinode.go:56 +0xb6
```

This is happening reliably with the Put acceptance test, by the way:
`TESTS=TestPut TESTTIMEOUT=24h TESTFLAGS="-d 600s -w" acceptance/run.sh`
after about 40-50k ops.
The first message in the log below is debug added by me in `range_raftstorage.go#Entries`. It likely has to do with the range split (range 12 below resulted from a split announced ~7sec earlier) and the new range winding up without a Raft log. So maybe some race, but it's happening pretty consistently beyond the 40k entries mark.

@bdarnell, any ideas?

>
W0608 14:16:39.880474       1 range_raftstorage.go:106] raft 12: can't load [0,1) with maxBytes 0, found:
[]
panic: requested entry at index is unavailable
goroutine 68 [running]:
github.com/coreos/etcd/raft.(*raft).sendAppend(0xc20807cd80, 0x100000001)
        /go/src/github.com/coreos/etcd/raft/raft.go:250 +0x159
github.com/coreos/etcd/raft.stepLeader(0xc20807cd80, 0x4, 0x300000003, 0x100000001, 0x6, 0x0, 0xa, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/coreos/etcd/raft/raft.go:523 +0x6f3
github.com/coreos/etcd/raft.(*raft).Step(0xc20807cd80, 0x4, 0x300000003, 0x100000001, 0x6, 0x0, 0xa, 0x0, 0x0, 0x0, ...)
        /go/src/github.com/coreos/etcd/raft/raft.go:487 +0x259
github.com/coreos/etcd/raft.(*multiNode).run(0xc2080be300)
        /go/src/github.com/coreos/etcd/raft/multinode.go:236 +0xd0b
created by github.com/coreos/etcd/raft.StartMultiNode
        /go/src/github.com/coreos/etcd/raft/multinode.go:56 +0xb6

--- a/storage/client_raft_test.go
+++ b/storage/client_raft_test.go
@@ -740,11 +740,6 @@ func TestRangeDescriptorSnapshotRace(t *testing.T) {
 	mtc := startMultiTestContext(t, 1)
 	defer mtc.Stop()
 
-	rng, err := mtc.stores[0].GetRange(1)
-	if err != nil {
-		t.Fatal(err)
-	}
-
 	stopper := util.NewStopper()
 	defer stopper.Stop()
 	// Call Snapshot() in a loop and ensure it never fails.
@@ -754,17 +749,51 @@ func TestRangeDescriptorSnapshotRace(t *testing.T) {
 			case <-stopper.ShouldStop():
 				return
 			default:
+				rng := mtc.stores[0].LookupRange(proto.KeyMin, nil)
+				if rng == nil {
+					t.Fatal("failed to look up min range")
+				}
 				_, err := rng.Snapshot()
 				if err != nil {
-					t.Fatal(err)
+					t.Fatalf("failed to snapshot min range: %s", err)
+				}
+
+				rng = mtc.stores[0].LookupRange(proto.Key("Z"), nil)
+				if rng == nil {
+					t.Fatal("failed to look up max range")
+				}
+				_, err = rng.Snapshot()
+				if err != nil {
+					t.Fatalf("failed to snapshot max range: %s", err)
 				}
 			}
 		}
 	})
 
-	// Split the range 100 times (carving chunks off the end of the initial range).
-	for i := 100; i > 0; i-- {
-		args, reply := adminSplitArgs(proto.KeyMin, []byte(fmt.Sprintf("%03d", i)), 1, mtc.stores[0].StoreID())
+	// Split the range repeatedly, carving chunks off the end of the
+	// initial range.  The bug that this test was designed to find
+	// usually occurred within the first 5 iterations.
+	for i := 20; i > 0; i-- {
+		rng := mtc.stores[0].LookupRange(proto.KeyMin, nil)
+		if rng == nil {
+			t.Fatal("failed to look up min range")
+		}
+		args, reply := adminSplitArgs(proto.KeyMin, []byte(fmt.Sprintf("A%03d", i)), rng.Desc().RaftID,
+			mtc.stores[0].StoreID())
+		rng.AdminSplit(args, reply)
+		if reply.GoError() != nil {
+			t.Fatal(reply.GoError())
+		}
+	}
+
+	// Split again, carving chunks off the beginning of the final range.
+	for i := 0; i < 20; i++ {
+		rng := mtc.stores[0].LookupRange(proto.Key("Z"), nil)
+		if rng == nil {
+			t.Fatal("failed to look up max range")
+		}
+		args, reply := adminSplitArgs(proto.KeyMin, []byte(fmt.Sprintf("B%03d", i)), rng.Desc().RaftID,
+			mtc.stores[0].StoreID())
 		rng.AdminSplit(args, reply)
 		if reply.GoError() != nil {
 			t.Fatal(reply.GoError())

--- a/storage/range_raftstorage.go
+++ b/storage/range_raftstorage.go
@@ -169,7 +169,12 @@ func (r *Range) FirstIndex() (uint64, error) {
 
 // loadAppliedIndex retrieves the applied index from the supplied engine.
 func (r *Range) loadAppliedIndex(eng engine.Engine) (uint64, error) {
-	appliedIndex := uint64(0)
+	var appliedIndex uint64
+	if r.isInitialized() {
+		appliedIndex = raftInitialLogIndex
+	} else {
+		appliedIndex = 0
+	}
 	v, err := engine.MVCCGet(eng, keys.RaftAppliedIndexKey(r.Desc().RaftID),
 		proto.ZeroTimestamp, true, nil)
 	if err != nil {
@@ -275,7 +280,7 @@ func (r *Range) Snapshot() (raftpb.Snapshot, error) {
 
 	term, err := r.Term(appliedIndex)
 	if err != nil {
-		return raftpb.Snapshot{}, err
+		return raftpb.Snapshot{}, util.Errorf("failed to fetch term of %d: %s", appliedIndex, err)
 	}
 
 	return raftpb.Snapshot{

